03-GettingStarted/06-http-streaming/README.md

```diff
@@ -0,0 +1,427 @@
+   - MCP: Progress is sent via separate notification messages while the main response comes at the end
+
+### Recommendations
+
+Based on what we've observed:
+
+1. **For simple streaming needs:** Classic HTTP streaming is simpler to implement and sufficient for basic streaming needs.
+
+2. **For complex, interactive applications:** MCP streaming provides a more structured approach with richer metadata and separation between notifications and final results.
+
+3. **For AI applications:** MCP's notification system is particularly useful for long-running AI tasks where you want to keep users informed of progress.
+
+---
+
+## 3. Streaming in MCP
+
+In MCP, streaming is not about sending the main response in chunks, but about sending **notifications** to the client while a tool is processing a request. These notifications can include progress updates, logs, or other events.
+
+#### Key differences from traditional streaming
+- The main result is still sent as a single response
+- Notifications are sent as separate messages during processing
+- The client must be able to handle and display these notifications
+
+#### What is a Notification?
+A notification is a message sent from the server to the client to inform about progress, status, or other events during a long-running operation. Notifications improve transparency and user experience.
+
+##### Why use notifications?
+- Keep users informed about progress
+- Enable real-time feedback and better UX
+- Useful for debugging and monitoring
+
+---
+
+## 4. Implementing Notifications in MCP
+
+### Server-side: Sending Notifications
+To send notifications from your MCP tool, use the context object (usually `ctx`) to call methods like `ctx.info()` or `ctx.log()`. These send messages to the client as the server processes a request.
+
+```python
+@mcp.tool(description="A tool that sends progress notifications")
+async def process_files(message: str, ctx: Context) -> TextContent:
+    await ctx.info("Processing file 1/3...")
+    await ctx.info("Processing file 2/3...")
+    await ctx.info("Processing file 3/3...")
+    return TextContent(type="text", text=f"Done: {message}")
+```
+
+### Client-side: Receiving Notifications
+The client must implement a message handler to process and display notifications as they arrive.
+
+```python
+async def message_handler(message):
+    if isinstance(message, types.ServerNotification):
+        print("NOTIFICATION:", message)
+    else:
+        print("SERVER MESSAGE:", message)
+```
+
+To enable notifications, ensure your server uses a streaming transport (like `streamable-http`) and your client implements a message handler to process notifications.
+
+---
+
+## 5. Streamable HTTP Transport
+
+### What is Streamable HTTP?
+Streamable HTTP is a transport mechanism in MCP that uses HTTP POST requests and supports streaming notifications (such as progress updates) from the server to the client. It is designed for modern web and cloud environments.
+
+### How It Works
```

Comment:
I suggest converting this bullet list to normal paragraphs.. Also please add a mermaid diagram so it's easier to understand the flow. People respond well to a visual